### DISCUSSION TITLE
Team : unable to extend a global role by a team

### DISCUSSION THREAD

**Discussion by Sean-OShea** [spatie/laravel-permission#1994] [Q&A] [answered]
Hi,

I've been searching for someone talking about that but i haven't found anything, so here is my use case. I have a set of global roles that i use as the default roles for creating a team. But the only difference between the global role and the one in the team would be the team id. The database allows that because the contraint enforces a unique composition of team_id/name/guard_name, but the `Spatie\Permission\Models\Role` still throws a `RoleAlreadyExists` because its `findByParam` method searches if the role already exists with or without a team_id : 

```
protected static function findByParam(array $params = [])
    {
        $query = static::when(PermissionRegistrar::$teams, function ($q) use ($params) {
            $q->where(function ($q) use ($params) {
                $q->whereNull(PermissionRegistrar::$teamsKey)
                    ->orWhere(PermissionRegistrar::$teamsKey, $params[PermissionRegistrar::$teamsKey] ?? app(PermissionRegistrar::class)->getPermissionsTeamId());
            });
        });
        unset($params[PermissionRegistrar::$teamsKey]);
        foreach ($params as $key => $value) {
            $query->where($key, $value);
        }

        return $query->first();
    }
```
Am i wrong in my understanding or is this an issue?
Thanks,
Sean.

**Comment by erikn69** [comment]
Global roles must be uniques by name and guard_name, if a composition of name/guard_name exists as global, you cant duplicate it on a team

**Comment by Sean-OShea** [comment]
Hi @erikn69, thanks for the answer. I see how it works but i can't find a justification of this behavior. There are only two places where that constraint is enforced in the code so i'm going to test if by extending those parts everything still works.

- I created my `Role` model that extends `Spatie\Permission\Models\Role` in order to override the `findByParam` function
- In my `User` model, I added a `role()` function in order to override the one in the `HasRole` trait.

The only change is that i removed the `whereNull` in both queries. I'll keep you updated of my tests today, i don't feel that it's the best way to do this but can't help to try it out :)

**Comment by Sean-OShea** [comment, marked as answer]
Found a more elegant solution so i don't have to extend spatie source code. I set a `team_id` to `0` on my "global" roles. Might answer all my needs :+1:
